src/buildstream/sandbox/_sandboxreapi.py

```diff
@@ -278,3 +278,10 @@ def execute_command(self, command):
 
     def create_empty_file(self, name):
         self.script += "touch -- {}\n".format(shlex.quote(name))
+
+    def clean_directory(self, name):
```

Comment:
This implementation appears to add shell commands, assuming that `rm` and `mkdir` will exist in the sandbox at runtime.

It looks like unfortunately we already went down this path with `touch` in `create_empty_file()` above.

My understanding is that *currently* we execute sandbox commands with `/bin/sh` by default, but there is nothing preventing us from working with other interpretors, or running commands in a sandbox that has no `/bin/sh` (i.e. we need only add some options to `BuildElement` to select what shell to use when invoking the sandbox).

Is having a shell a new limitation stemming from remote execution ? Shouldn't we have some way of mutating the remote sandbox content (adding/removing files/directories) without executing commands ?

Reply:
`mkdir` is already required to ensure the working directory exists a bit further up. Expecting `rm` to exist is not a big stretch. Also, the build will not fail if `rm` doesn't exist, so this optimization shouldn't break anything.

Yes, using a shell script was part of the switch to REAPI, initially only for remote execution and later also for buildbox-run. With remote execution we can only execute a single command in a sandbox but that can be `sh`. We can manipulate the input tree beforehand in any way, of course, but can't manipulate the sandbox while remote execution is already running. Issuing individual commands using separate sandboxes / REAPI executions would be terrible for performance.

Sandbox support without POSIX `sh` is certainly worth discussing. Configuration at the point of `sandbox.batch()` may make sense. However, I don't think this branch makes anything worse as it generates commands in the same place (`SandboxREAPI`) as what's already in master and doesn't go beyond very basic POSIX commands.